### ISSUE TITLE
[no-unnecessary-condition] rule reports even when the condition is necessary using noUncheckedIndexedAccess

### ISSUE THREAD

**Issue by dev054** [typescript-eslint/typescript-eslint#2966]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
With `noUncheckedIndexedAccess` enabled in tsconfig.json, I have this:

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": true
  }
}
```

```TS
const items = [{ a: 1 }, { a: 2 }] as const;
const item = items[index]; // index is something that come dynamically
```

`item` is: 
```TS
const item: {
    readonly a: 1;
} | {
    readonly a: 2;
} | undefined
```

Now I must check the nullability of item before using it as it can be `undefined`:
```TS
if (!item) return;
```

... however I get:
> Unnecessary conditional, value is always falsy. eslint@typescript-eslint/no-unnecessary-condition

<hr>

Another case is this:

```TS
const pattern = /_([a-z\d])/g;

return someString.replace(
  pattern,
  // Here it reports, saying that it's an unnecessary condition.
  ({ 1: firstLetter }) => firstLetter?.toUpperCase() ?? '',
);
```

**Expected Result**
I would expect no reports.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.0` |
| `@typescript-eslint/parser`        | `4.14.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.18.0` |

**Comment by bradzacher** [comment]
Could you please provide your tsconfig?

**Comment by dev054** [comment]
After reinstalling everything from scratch the erros are gone :man_shrugging:. Sorry for that.
